### ISSUE TITLE
UserStackFrame wrong when logger wrapped in composition

### ISSUE THREAD

**Issue by JueSte** [NLog/NLog#1070]
Our wrapper uses a composition and it works fine with Nlog 3.1
I think this is related to #541

This unit test should explain the issue:

```c#
[Fact]
public void When_WrappedInCompsition_Ignore_Wrapper_Methods_In_Callstack()
{

    //namespace en name of current method
    const string currentMethodFullName = "NLog.UnitTests.LayoutRenderers.CallSiteTests.When_WrappedInCompsition_Ignore_Wrapper_Methods_In_Callstack";

    LogManager.Configuration = CreateConfigurationFromString(@"
       <nlog>
           <targets><target name='debug' type='Debug' layout='${callsite}|${message}' /></targets>
           <rules>
               <logger name='*' levels='Warn' writeTo='debug' />
           </rules>
       </nlog>");

    var logger = LogManager.GetLogger("A");
    logger.Warn("direct");
    AssertDebugLastMessage("debug", string.Format("{0}|direct", currentMethodFullName));

    LoggerTests.CompositeWrapper wrappedLogger = new LoggerTests.CompositeWrapper();
    wrappedLogger.Log("wrapped");
    AssertDebugLastMessage("debug", string.Format("{0}|wrapped", currentMethodFullName));
    
}



public class CompositeWrapper
{
    private readonly MyWrapper wrappedLogger;
    public CompositeWrapper()
    {
        wrappedLogger = new MyWrapper();
    }
    public void Log(string what)
    {
        wrappedLogger.Log(typeof(CompositeWrapper), what);
    }
}

public abstract class BaseWrapper
{
    public void Log(string what)
    {
        InternalLog(typeof(BaseWrapper), what);
    }

    public void Log(Type type, string what) //overloaded with type for composition
    {
        InternalLog(type, what);
    }

    protected abstract void InternalLog(Type type,string what);
}

public class MyWrapper : BaseWrapper
{
    private readonly ILogger wrapperLogger;

    public MyWrapper()
    {
        wrapperLogger = LogManager.GetLogger("WrappedLogger");
    }

    protected override void InternalLog(Type type, string what) //added type for composition
    {
        LogEventInfo info = new LogEventInfo(LogLevel.Warn, wrapperLogger.Name, what);

        // Provide BaseWrapper as wrapper type.
        // Expected: UserStackFrame should point to the method that calls a 
        // method of BaseWrapper.
        wrapperLogger.Log(type, info);
    }
}
```

**Comment by 304NotModified** [comment]
Thanks for the unit test!

I will test this in the latest version. 



> I think this is related to #541

This one has been fixed in 4.0

**Comment by 304NotModified** [comment]
- `When_Wrapped_Ignore_Wrapper_Methods_In_Callstack`: succeeds
- `When_WrappedInCompsition_Ignore_Wrapper_Methods_In_Callstack`: fails (second assert)

stacktrace 1:
```
   at NLog.LoggerImpl.Write(Type loggerType, TargetWithFilterChain targets, LogEventInfo logEvent, LogFactory factory)
   at NLog.Logger.WriteToTargets(LogLevel level, IFormatProvider formatProvider, String message)
   at NLog.Logger.Warn(String message)
   at NLog.UnitTests.LayoutRenderers.CallSiteTests.When_WrappedInCompsition_Ignore_Wrapper_Methods_In_Callstack()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Xunit.Sdk.Reflector.ReflectionMethodInfo.Invoke(Object testClass, Object[] parameters)
   at Xunit.Sdk.FactCommand.Execute(Object testClass)
   at Xunit.Sdk.FixtureCommand.Execute(Object testClass)
   at Xunit.Sdk.BeforeAfterCommand.Execute(Object testClass)
   at Xunit.Sdk.LifetimeCommand.Execute(Object testClass)
   at Xunit.Sdk.ExceptionAndOutputCaptureCommand.Execute(Object testClass)
   at Xunit.Sdk.TimedCommand.Execute(Object testClass)
   at Xunit.Sdk.TestClassCommandRunner.Execute(ITestClassCommand testClassCommand, List`1 methods, Predicate`1 startCallback, Predicate`1 resultCallback)
   at Xunit.Sdk.Executor.RunTests.<>c__DisplayClass12.<.ctor>b__f()
   at Xunit.Sdk.Executor.ThreadRunner(Object threadStart)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)

```
stacktrace 2:

```
   at NLog.LoggerImpl.Write(Type loggerType, TargetWithFilterChain targets, LogEventInfo logEvent, LogFactory factory)
   at NLog.Logger.WriteToTargets(Type wrapperType, LogEventInfo logEvent)
   at NLog.Logger.Log(Type wrapperType, LogEventInfo logEvent)
   at NLog.UnitTests.LayoutRenderers.CallSiteTests.MyWrapper.InternalLog(Type type, String what)
   at NLog.UnitTests.LayoutRenderers.CallSiteTests.BaseWrapper.Log(Type type, String what)
   at NLog.UnitTests.LayoutRenderers.CallSiteTests.CompositeWrapper.Log(String what)
   at NLog.UnitTests.LayoutRenderers.CallSiteTests.When_WrappedInCompsition_Ignore_Wrapper_Methods_In_Callstack()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Xunit.Sdk.Reflector.ReflectionMethodInfo.Invoke(Object testClass, Object[] parameters)
   at Xunit.Sdk.FactCommand.Execute(Object testClass)
   at Xunit.Sdk.FixtureCommand.Execute(Object testClass)
   at Xunit.Sdk.BeforeAfterCommand.Execute(Object testClass)
   at Xunit.Sdk.LifetimeCommand.Execute(Object testClass)
   at Xunit.Sdk.ExceptionAndOutputCaptureCommand.Execute(Object testClass)
   at Xunit.Sdk.TimedCommand.Execute(Object testClass)
   at Xunit.Sdk.TestClassCommandRunner.Execute(ITestClassCommand testClassCommand, List`1 methods, Predicate`1 startCallback, Predicate`1 resultCallback)
   at Xunit.Sdk.Executor.RunTests.<>c__DisplayClass12.<.ctor>b__f()
   at Xunit.Sdk.Executor.ThreadRunner(Object threadStart)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)


```

**Comment by 304NotModified** [comment]
difficult one, as we now record the last call in user code (`MyWrapper.InternalLog`), while in the test case the first call is expected (`CallSiteTests.When_WrappedInCompsition_Ignore`).

I have a fix that succeeds on both tests, but other unit test fail then. 

So I think so that the expectations are incorrect. Do you agree? The first call can never work as this the `main()` method in a console program.

**Comment by JueSte** [comment]
I can do a workaround with `LogManager.AddHiddenAssembly()` and our wrapper will continue to work. Thank you for this tip. 

However my expectation how the `wrapperType` parameter works is this:
I pass the type of the first class which counts as `logger` for me and I don't want any method calls that appear in the call stack after this class influence the call site log entry.
I don't expect that it has to be a base type of my Logger encapsulation.

So I don't agree that these unit test is a false expectation. 

I think the `IgnoreNamespaces` would be a good solution to meet the requirement that I can wrap the logger in any way i like, but it wouldn't fix the unexpected behavior of `wrapperType`.

**Comment by 304NotModified** [comment]
> However my expectation how the wrapperType parameter works is this:
> I pass the type of the first class which counts as logger for me and I don't want any method calls that appear in the call stack after this class influence the call site log entry.

I this case it unit test (2nd call) it is `CompositeWrapper` ? It should be the fist call after the wrappertype? (maybe you are right, not sure)

```
   * at MyWrapper.InternalLog(Type type, String what) <--------- choosen (wrong)
   * at BaseWrapper.Log(Type type, String what)
   * at CompositeWrapper.Log(String what) <!-- type of wrapperType
   * at When_WrappedInCompsition_Ignore_Wrapper_Methods_In_Callstack() <-- expected

```

**Comment by JueSte** [comment]
Yes that is what i would expect.

**Comment by 304NotModified** [comment]
OK Thanks! :) 

I will check this again, probably this weekend.

**Comment by 304NotModified** [comment]
hey!'

It took a bit more them, and I re-implemented the code. I got it fixed, but I'm not sure if it's fully backwards-compatible / 100% what is expected.

The algorithm is now:

1. Find LoggerType
2. Take next
3. Skip if in SkipAssembly

Before (the code wasn't really readable), it checked for "is loggertype or IsAssignableFrom loggertype". This is now removed, otherwise the submitted case won't work. Do you think that's correct?

**Comment by 304NotModified** [comment]
This is fixed in 4.3. See #1095.
